File: src/workers/claude-code.ts
import type { WorkerAdapter, WorkerEvent, WorkerResult, WorkerSpawnOptions } from '../types.js';
import type { SpawnCliResult } from './spawn-utils.js';
import { runWorkerCli } from './herdr-runtime.js';
import { prepareWorkerSpawn } from './sandbox.js';
import { backoffDelay } from '../sessionguard/backoff.js';
import { isNonRetryableApiError } from '../sessionguard/events.js';
import { isRetryableWithoutSession } from '../resilience/classify.js';

const RESUME_PROMPT = 'Continue';
const DEFAULT_API_MAX_ATTEMPTS = Infinity;

function resolveNoProgressTimeoutMs(explicit: number | undefined): number {
  if (explicit !== undefined) return explicit;
  const env = process.env.DEVAGENT_NO_PROGRESS_TIMEOUT_MS;
  if (env !== undefined && env !== '') {
    const n = Number(env);
    if (Number.isFinite(n) && n >= 0) return n;
  }
  // 0 = watchdog off (callers like deps.ts/consume use config to pass 10m in prod).
  // Keeping direct adapter invocations fast in unit tests.
  return 0;
}

/**
 * Adapter over the Claude Code headless CLI:
 *   claude -p <prompt> --output-format json [--max-turns N]
 * stdout is a single JSON object with fields like `result` and `session_id`.
 *
 * Claude Code never retries a response dropped mid-stream; when that happens
 * the turn dies but the session transcript persists. This adapter resumes the
 * same session (`--resume <id> -p "Continue"`) with exponential backoff until
 * the turn completes, attempts are exhausted, or the error is auth/billing.
 */
export class ClaudeCodeAdapter implements WorkerAdapter {
  readonly name = 'claude-code' as const;

  constructor(
    private readonly sleep: (ms: number) => Promise<void> = (ms) =>
      new Promise((resolve) => setTimeout(resolve, ms)),
  ) {}

  async spawn(opts: WorkerSpawnOptions): Promise<WorkerResult> {
    const start = Date.now();
    const maxAttempts = opts.apiMaxAttempts ?? DEFAULT_API_MAX_ATTEMPTS;
    const noProgressTimeoutMs = resolveNoProgressTimeoutMs(opts.noProgressTimeoutMs);
    const wallDeadline = opts.timeoutMs > 0 ? start + opts.timeoutMs : Infinity;

    let args = baseArgs(opts);
    let sessionId: string | null = null;
    let last: SpawnCliResult | null = null;

    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      // Wall-clock budget: stop retrying when the run's overall timeout is spent.
      if (Date.now() >= wallDeadline) {
        if (last) last.timedOut = true;
        break;
      }
      const prepared = await prepareWorkerSpawn('claude', args, {
        cwd: opts.cwd,
        timeoutMs: opts.timeoutMs,
        ...(opts.env ? { env: opts.env } : {}),
        ...(noProgressTimeoutMs ? { noProgressTimeoutMs } : {}),
        ...(opts.watchdogLedger ? { watchdogLedger: opts.watchdogLedger } : {}),
      });
      last = await runWorkerCli(prepared.cmd, prepared.args, {
        ...prepared.opts,
        ...(noProgressTimeoutMs ? { noProgressTimeoutMs } : {}),
        ...(opts.herdr ? { herdr: true } : {}),
      });

      const outcome = interpret(last);
      if (outcome.sessionId) sessionId = outcome.sessionId;

      const ok = !last.timedOut && outcome.exitCode === 0 && !outcome.isError;
      if (ok) break;

      if (attempt === maxAttempts) break;
      if (last.exitCode === -1 && !last.timedOut) break; // spawn failure (ENOENT) — never retry forever
      if (outcome.errorText && isNonRetryableApiError(outcome.errorText)) break;

      // Transient detection: watchdog timeouts and provider errors (Console
      // Go, upstream, etc.) are retried forever. Without a session we retry
      // from scratch only for provider/timeout signals (not generic ECONNREFUSED).
      const errorText = outcome.errorText ?? last.stderr ?? '';

      if (!sessionId) {
        const retryableWithoutSession = isRetryableWithoutSession({
          timedOut: last.timedOut,
          exitCode: last.exitCode,
          errorText,
          stderr: last.stderr,
        });
        if (!retryableWithoutSession) break;
      }
      // Wall-clock budget already checked above; also don't retry if we'd exceed it after sleep
      if (Date.now() >= wallDeadline) break;

      await this.sleep(backoffDelay(attempt));
      if (sessionId) {
        args = [
          '--resume',
          sessionId,
          '-p',
          RESUME_PROMPT,
          '--output-format',
          'json',
          ...(opts.maxSteps !== undefined ? ['--max-turns', String(opts.maxSteps)] : []),
        ];
      } else {
        args = baseArgs(opts);
      }
    }

    return finalize(last!, start);
  }
}

function baseArgs(opts: WorkerSpawnOptions): string[] {
  // claude-code ignores variant; model is the only knob.
  // Driver tier aliases ("coding", devagent.json model) are claude-proxy
  // selectors, not API-key model ids: `--model coding` fails with
  // "403 Combo "coding" is not allowed for this API key" (2026-09-01 live
  // loop 58: implement worker retried it forever because events=0 +
  // null resultText reads as a logic failure, not transient). Drop any
  // value without a "/" or known claude family prefix so claude falls back
  // to ~/.claude/settings.json model. Pass explicit provider ids through.
  const rawModel = opts.model?.trim();
  const model =
    rawModel && (/^claude-/.test(rawModel) || /^(opus|sonnet|haiku)(-|$)/.test(rawModel))
      ? rawModel.split('#')[0]!
      : undefined;
  return [
    '-p',
    opts.prompt,
    '--output-format',
    'json',
    ...(model ? ['--model', model] : []),
    ...(opts.maxSteps !== undefined ? ['--max-turns', String(opts.maxSteps)] : []),
  ];
}

interface RunOutcome {
  exitCode: number;
  isError: boolean;
  sessionId: string | null;
  errorText?: string;
  parsed: Record<string, unknown> | null;
}

export function interpretForTest(run: SpawnCliResult): RunOutcome {
  return interpret(run);
}

function interpret(run: SpawnCliResult): RunOutcome {
  let parsed: Record<string, unknown> | null = null;
  if (run.stdout.trim()) {
    try {
      const candidate: unknown = JSON.parse(run.stdout);
      if (Array.isArray(candidate)) {
        // Newer claude CLI: --output-format json emits the full event stream
        // as a JSON array; the terminal entry carries type:'result'
        // (live-smoke lesson: treating arrays as unparsable made every
        // planner call look empty).
        for (let i = candidate.length - 1; i >= 0; i--) {
          const e = candidate[i] as Record<string, unknown> | null;
          if (e !== null && typeof e === 'object' && !Array.isArray(e) && e.type === 'result') {
            parsed = e;
            break;
          }
        }
      } else if (candidate !== null && typeof candidate === 'object') {
        parsed = candidate as Record<string, unknown>;
      }
    } catch {
      parsed = null;
    }
  }
  const isError = parsed?.is_error === true;
  const rawResult = parsed?.result;
  const sessionId =
    parsed !== null && typeof parsed.session_id === 'string' ? parsed.session_id : null;
  return {
    exitCode: run.exitCode,
    isError: isError || (run.exitCode !== 0 && parsed === null),
    sessionId,
    errorText:
      typeof rawResult === 'string' && rawResult
        ? rawResult
        : run.stderr.trim()
          ? run.stderr.trim().split('\n').slice(-3).join('\n')
          : undefined,
    parsed,
  };
}

function finalize(run: SpawnCliResult, start: number): WorkerResult {
  if (run.timedOut) {
    return {
      exitCode: run.exitCode,
      events: [],
      resultText: null,
      sessionId: null,
      durationMs: Date.now() - start,
      timedOut: true,
      errorText: run.stderr.trim() || undefined,
    };
  }

  const outcome = interpret(run);
  if (!run.stdout.trim() || outcome.parsed === null) {
    return {
      exitCode: outcome.exitCode,
      events: [],
      resultText: null,
      sessionId: outcome.sessionId,
      durationMs: Date.now() - start,
      timedOut: false,
      // Stderr often carries the real failure reason (e.g.
      // [claude-code:unrecognized_model]) when the upstream returned an
      // empty stream. Surface it so the executor can classify properly.
      errorText: outcome.errorText ?? (run.stderr.trim() || undefined),
    };
  }

  const events: WorkerEvent[] = [{ type: 'result', ...outcome.parsed }];
  const resultText =
    outcome.exitCode === 0 && typeof outcome.parsed.result === 'string'
      ? outcome.parsed.result
      : null;

  return {
    exitCode: outcome.exitCode,
    events,
    resultText,
    sessionId: outcome.sessionId,
    durationMs: Date.now() - start,
    timedOut: false,
    errorText: outcome.errorText,
  };
}

File: internal/workers/claudecode.go
// Go port of src/workers/claude-code.ts (FR-GO-05).
//
// Adapter over the Claude Code headless CLI:
//
//	claude -p <prompt> --output-format json [--max-turns N]
//
// stdout is a single JSON object with fields like `result` and
// `session_id` (newer CLIs emit the full event stream as a JSON array; the
// terminal entry carries type:'result').
//
// Claude Code never retries a response dropped mid-stream; when that
// happens the turn dies but the session transcript persists. This adapter
// resumes the same session (`--resume <id> -p "Continue"`) with exponential
// backoff until the turn completes, attempts are exhausted, or the error is
// auth/billing.

package workers

import (
	"regexp"
	"strings"
)

const claudeResumePrompt = "Continue"

// TS Infinity proxy: effectively unbounded attempt budget.
const claudeDefaultAPIMaxAttempts = 1 << 30

// ClaudeCodeAdapter mirrors TS ClaudeCodeAdapter. The embedded
// adapterBase carries the test seams (Sleep/Run/Prepare/NowMs); nil =
// production behavior.
type ClaudeCodeAdapter struct {
	adapterBase
}

// Name mirrors the TS readonly name.
func (a *ClaudeCodeAdapter) Name() string { return "claude-code" }

// Capabilities: Q30 — declares no watchdog budget (0 = watchdog off;
// callers like the executor arm one explicitly from config when they want
// it).
func (a *ClaudeCodeAdapter) Capabilities() WorkerCapabilities {
	return WorkerCapabilities{DefaultNoProgressTimeoutMs: 0}
}

// IsProgress: claude-code emits single-JSON envelopes rather than an
// NDJSON event stream; TS leaves isProgress undefined for this adapter, so
// the runtime fallback keeps the shared NDJSON core semantics.
func (a *ClaudeCodeAdapter) IsProgress(line string) bool { return IsNdjsonProgressLine(line) }

// Spawn runs the adapter retry loop.
func (a *ClaudeCodeAdapter) Spawn(opts WorkerSpawnOptions) WorkerResult {
	start := a.nowMs()
	maxAttempts := claudeDefaultAPIMaxAttempts
	if opts.APIMaxAttempts != nil {
		maxAttempts = *opts.APIMaxAttempts
	}
	caps := a.Capabilities()
	noProgressTimeoutMs := ResolveNoProgressTimeoutMs(opts.NoProgressTimeoutMs, &caps)
	wallDeadline := int64(1 << 62)
	if opts.TimeoutMs > 0 {
		wallDeadline = start + int64(opts.TimeoutMs)
	}

	args := claudeBaseArgs(opts)
	sessionId := ""
	var last *SpawnCliResult

	for attempt := 1; attempt <= maxAttempts; attempt++ {
		// Wall-clock budget: stop retrying when the run's overall timeout
		// is spent.
		if a.nowMs() >= wallDeadline {
			if last != nil {
				last.TimedOut = true
			}
			break
		}
		spawnOpts := SpawnCliOptions{
			Dir:                 opts.Cwd,
			TimeoutMs:           opts.TimeoutMs,
			Env:                 opts.Env,
			NoProgressTimeoutMs: intPtrIf(noProgressTimeoutMs != 0, noProgressTimeoutMs),
			ColdStartTimeoutMs:  opts.ColdStartTimeoutMs,
			WatchdogLedger:      opts.WatchdogLedger,
		}
		prepared, err := a.prepare("claude", args, spawnOpts)
		if err != nil {
			// prepareWorkerSpawn failures are loud by contract; surface as
			// a spawn failure result instead of panicking the dispatcher.
			res := SpawnCliResult{ExitCode: -1, Stderr: err.Error()}
			last = &res
			break
		}
		runOpts := RunWorkerCliOptions{SpawnCliOptions: prepared.Opts, Herdr: opts.Herdr}
		raw := a.run(prepared.Cmd, prepared.Args, runOpts)
		last = &raw

		outcome := claudeInterpret(raw)
		if outcome.sessionId != "" {
			sessionId = outcome.sessionId
		}

		ok := !raw.TimedOut && outcome.exitCode == 0 && !outcome.isError
		if ok {
			break
		}

		if attempt == maxAttempts {
			break
		}
		if raw.ExitCode == -1 && !raw.TimedOut {
			break // spawn failure (ENOENT) — never retry forever
		}
		if outcome.errorText != "" && IsNonRetryableApiError(outcome.errorText) {
			break
		}

		// Transient detection: watchdog timeouts and provider errors
		// (Console Go, upstream, etc.) are retried forever. Without a
		// session we retry from scratch only for provider/timeout signals
		// (not generic ECONNREFUSED).
		errorText := outcome.errorText
		if errorText == "" {
			errorText = raw.Stderr
		}
		if sessionId == "" {
			if !IsRetryableWithoutSession(raw.TimedOut, raw.ExitCode, errorText, raw.Stderr) {
				break
			}
		}
		// Wall-clock budget already checked above; also don't retry if
		// we'd exceed it after sleep.
		if a.nowMs() >= wallDeadline {
			break
		}

		a.sleepMs(backoffDelay(attempt))
		if sessionId != "" {
			resume := []string{
				"--resume",
				sessionId,
				"-p",
				claudeResumePrompt,
				"--output-format",
				"json",
			}
			if opts.MaxSteps != nil {
				resume = append(resume, "--max-turns", itoa(*opts.MaxSteps))
			}
			args = resume
		} else {
			args = claudeBaseArgs(opts)
		}
	}

	return claudeFinalize(derefSpawn(last), a.nowMs()-start)
}

// claudeBaseArgs mirrors the TS baseArgs. claude-code ignores variant;
// model is the only knob. Driver tier aliases ("coding", devagent.json
// model) are claude-proxy selectors, not API-key model ids: `--model
// coding` fails with "403 Combo "coding" is not allowed for this API key"
// (2026-09-01 live loop 58: implement worker retried it forever because
// events=0 + null resultText reads as a logic failure, not transient).
// Drop any value without a "/" or known claude family prefix so claude
// falls back to ~/.claude/settings.json model. Pass explicit provider ids
// through.
func claudeBaseArgs(opts WorkerSpawnOptions) []string {
	rawModel := strings.TrimSpace(opts.Model)
	var model string
	if rawModel != "" && claudeModelRe.MatchString(rawModel) {
		// TS rawModel.split('#')[0]
		if i := strings.IndexByte(rawModel, '#'); i >= 0 {
			model = rawModel[:i]
		} else {
			model = rawModel
		}
	}
	args := []string{"-p", opts.Prompt, "--output-format", "json"}
	if model != "" {
		args = append(args, "--model", model)
	}
	if opts.MaxSteps != nil {
		args = append(args, "--max-turns", itoa(*opts.MaxSteps))
	}
	return args
}

// claudeModelRe mirrors the TS regex pair
// /^claude-/.test(rawModel) || /^(opus|sonnet|haiku)(-|$)/.test(rawModel).
var claudeModelRe = regexp.MustCompile(`^claude-|^(?:opus|sonnet|haiku)(?:-|$)`)

// claudeRunOutcome mirrors TS RunOutcome.
type claudeRunOutcome struct {
	exitCode  int
	isError   bool
	sessionId string
	errorText string
	parsed    map[string]any
}

// InterpretClaudeForTest is the test seam re-export of the parser.
func InterpretClaudeForTest(run SpawnCliResult) claudeRunOutcome {
	return claudeInterpret(run)
}

// claudeInterpret parses claude's stdout into the worker outcome shape.
func claudeInterpret(run SpawnCliResult) claudeRunOutcome {
	var parsed map[string]any
	if s := strings.TrimSpace(run.Stdout); s != "" {
		if candidate, ok := parseJSONAny(s); ok {
			switch v := candidate.(type) {
			case []any:
				// Newer claude CLI: --output-format json emits the full
				// event stream as a JSON array; the terminal entry carries
				// type:'result' (live-smoke lesson: treating arrays as
				// unparsable made every planner call look empty).
				for i := len(v) - 1; i >= 0; i-- {
					if e, ok := v[i].(map[string]any); ok && e["type"] == "result" {
						parsed = e
						break
					}
				}
			case map[string]any:
				parsed = v
			}
		}
	}
	isError := asBool(parsed["is_error"])
	rawResult, _ := parsed["result"].(string)
	sessionId := ""
	if parsed != nil {
		sessionId, _ = parsed["session_id"].(string)
	}
	errorText := ""
	if rawResult != "" {
		errorText = rawResult
	} else if t := strings.TrimSpace(run.Stderr); t != "" {
		// TS: stderr.trim().split('\n').slice(-3).join('\n')
		errorText = lastNLines(t, 3)
	}
	isError = isError || (run.ExitCode != 0 && parsed == nil)
	return claudeRunOutcome{
		exitCode:  run.ExitCode,
		isError:   isError,
		sessionId: sessionId,
		errorText: errorText,
		parsed:    parsed,
	}
}

// claudeFinalize mirrors TS finalize.
func claudeFinalize(run SpawnCliResult, durationMs int64) WorkerResult {
	if run.TimedOut {
		result := WorkerResult{
			ExitCode:   run.ExitCode,
			Events:     []WorkerEvent{},
			ResultText: "",
			SessionId:  "",
			DurationMs: durationMs,
			TimedOut:   true,
			ErrorText:  strings.TrimSpace(run.Stderr),
		}
		if run.ColdStart {
			result.ColdStart = true
		}
		return result
	}

	outcome := claudeInterpret(run)
	if strings.TrimSpace(run.Stdout) == "" || outcome.parsed == nil {
		// Stderr often carries the real failure reason (e.g.
		// [claude-code:unrecognized_model]) when the upstream returned an
		// empty stream. Surface it so the executor can classify properly.
		errorText := outcome.errorText
		if errorText == "" {
			errorText = strings.TrimSpace(run.Stderr)
		}
		return WorkerResult{
			ExitCode:   outcome.exitCode,
			Events:     []WorkerEvent{},
			ResultText: "",
			SessionId:  outcome.sessionId,
			DurationMs: durationMs,
			TimedOut:   false,
			ErrorText:  errorText,
		}
	}

	events := []WorkerEvent{eventFromResult(outcome.parsed)}
	resultText := ""
	if outcome.exitCode == 0 {
		if s, ok := outcome.parsed["result"].(string); ok {
			resultText = s
		}
	}
	return WorkerResult{
		ExitCode:   outcome.exitCode,
		Events:     events,
		ResultText: resultText,
		SessionId:  outcome.sessionId,
		DurationMs: durationMs,
		TimedOut:   false,
		ErrorText:  outcome.errorText,
	}
}
